### ISSUE TITLE
(Maybe) Split parsers.py into files

### ISSUE THREAD

**Issue by jmmelko** [sky-music/sky-python-music-sheet-maker#100]
As suggested by Strawb:

The code in parsers.py is very long.

The reason for that is that it handles many different input formats (aka song notations), and uses long hard coded variables, such as dictionaries, regular expressions, etc.

I think we could easily split the NoteParsers into different files:

* SkyKeyboardNoteParser
* SkyNoteParser
* EnglishNoteParser
* EnglishChordsNoteParser
* DoremiNoteParser
* JianpuNoteParser

I suppose we would also have to put the generic NoteParser in a separate file.

I can do it but I need confirmation from our Code Mistress.

This is our 100th issue!

Happy birthday everyone 🥳🎂

**Comment by t1-tracey** [comment]
😂 im ok with it. yeah each of the files will just have to import the generic NoteParser I think. Maybe they can be put into a `parsers` folder (would need an empty __init__.py in it)

**Comment by jmmelko** [comment]
Ok so I managed to do it (on my local files), here is what I did:

* python
     ** noteparsers 
           - __init__.py
           -noteparser.py (NoteParser)
           -doremi.py (Doremi)
           - etc

Each parser file contains the following code:
import os, sys
sys.path.append(os.path.dirname(os.path.abspath(__file__)))

so that they can find NoteParser which is in the same directory and not in the working directory (which is one level up).

I have simply named parsers English, Jianpu, etc, but I can add NoteParser if preferred.

Parsers and Song now contains the following imports:

from noteparsers.skykeyboard import SkyKeyboard
from noteparsers.sky import Sky
from noteparsers.english import English
from noteparsers.jianpu import Jianpu
from noteparsers.doremi import Doremi
from noteparsers.englishchords import EnglishChords

I had a problem with NoteParser and English because they are cross referencing each other. More specifically, NoteParser needs English for the english_note_name method.
I solved this by importing English only when needed:

    def english_note_name(self, note_name, reverse=False):
        from english import English


Please let me know if it’s ok.

I’d like to export the code somewhere for Tracey to check it before overwriting dev.

Create a temporary branch?

Thank you.

**Comment by t1-tracey** [comment]
Wow you’ve split them now, I think this is awesome work 😊 I think a new temporary branch off dev is ok ^^

**Comment by jmmelko** [comment]
Ok, I have created the branch dev-split.

The ideas is to use review my code, propose modifications, and then I will upload the final version in dev. I will ask Strawb to take a look.

**Comment by IchigoDaifuku** [comment]
Yep! This looks good!
Also, yes there is a way to simplify imports - you can put in the __init__.py file 
`__all__ = [
    "SkyKeyboard", 
    "Sky", 
    "English", 
    "Jianpy", 
    "Doremi", 
    "EnglishChords"
]`
And this will allow you to use
`from noteparsers import *`
but i'm sure this isn't pythonic and i do not know if its possible to make it more specific so you can import the classes in the module

**Comment by jmmelko** [comment]
Ok, I have made the change.
It is not pythonic in the sense that “explicit is better than implicit”. And importing all (“*”) is not recommended.

A more pythonic way to do that would be to simply import noteparsers (the folder as a package), and then instanciate objects with noteparsers.Jianpu.jianpu()
But this leads to longer call lines, and for some reason Tracey seldom uses this explicit embedded notation.

Tell vote for your final choice and I will then apply it to dev directly.

**Comment by t1-tracey** [comment]
> A more pythonic way to do that would be to simply import noteparsers (the folder as a package), and then instanciate objects with noteparsers.Jianpu.jianpu()

> But this leads to longer call lines, and for some reason Tracey seldom uses this explicit embedded notation.

I didn’t know about what’s best for instantiating objects e.g. with `noteparsers.Jianpu.jianpu()`. I was angry for a moment, XD I would like the issue to be raised if you feel there's a better practice for approaching something, and we can update the code accordingly (and I’ll follow it in future).

**Comment by IchigoDaifuku** [comment]
The note parsers only need to be instanciated once so using `noteparsers.Jianpu.jianpu()` would work. Longer call lines shouldn't be a problem if we are only using it once.

**Comment by t1-tracey** [comment]
Ok 😊 we can use it going forward, and fixing the previous usage of it.

**Comment by jmmelko** [comment]
> I was angry for a moment, XD I would like to be just informed if you feel there's a better practice for approaching something, and I’ll follow it in future.

Sorry, I think that’s because I am non native speaker.
To me, « for some reason » did *not* mean « she should know better », but more « for a good reason I a not aware of because I have a limited knowledge of the code ».

I think the code is short enough so we don’t have to brood over good practice and PEP8 too much, just for the sake of looking smart :-)

**Comment by jmmelko** [comment]
LOL, believe it or not, but there is no built-in function in Python to import all modules in a package (that is, to import all *.py from a folder).

So I will stick to import all.

Another way would be to write a function, but I think it is overkill:

https://stackoverflow.com/questions/3365740/how-to-import-all-submodules

**Comment by t1-tracey** [comment]
Aw it’s ok 🤗 Sorry I assumed too quickly. (Et je peux pas parler en français bien lol. )

**Comment by jmmelko** [comment]
Ok, so here is what I did finally:

* used explicit call of noteparsers, as advised by 🍓: noteparsers.jianpu.Jianpu()
* to do that, I had to write content in the \_\_init\_\_.py file of the noteparsers module
* I have kept the definition of _\_\all_\_\, it doesn’t hurt, we don’t have to use it
* I have cleaned up a bit import in other files, following this logic:
   - importing the whole module when several of its classes are used (for instance notes)
  - importing the class directly when it is used many times and when other classes are not useful for this module (e.g InputMode)
  - except for standard python packages (io, re, os, sys)
 - importing the class directly when it is alone in its module  (Song)


Anyway, we can change it later. I’m done with it. Sorry for taking your time.
